1. Завантажити біблітеку PyGad

In [2]:
!pip install PyGad

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 3.5 MB/s eta 0:00:00


In [3]:
import pygad
import numpy as np


2. Визначити окремо в коді фітнес функцію.

In [7]:
# Опис продуктів (назва, об'єм, ціна)
products = [
    ("TV 55", 0.400, 4346.99),
    ("TV 50", 0.290, 3999.90),
    ("TV 42", 0.200, 2999.00),
    ("Notebook A", 0.0035, 2499.90),
    ("Ventilator", 0.496, 199.90),
    ("Microwave A", 0.0424, 308.66),
    ("Microwave B", 0.0544, 429.90),
    ("Microwave C", 0.0319, 299.29),
    ("Refrigerator B", 0.635, 849.00),
    ("Refrigerator C", 0.870, 1199.89),
    ("Notebook B", 0.498, 1999.90),
    ("Notebook C", 0.527, 3999.00)
]

volumes = np.array([p[1] for p in products])
prices = np.array([p[2] for p in products])
max_volume = 3.0  # Максимальний об'єм вантажівки

# Фітнес-функція
def fitness_func(ga_instance, solution, solution_idx):
    total_volume = np.sum(solution * volumes)
    total_price = np.sum(solution * prices)
    if total_volume > max_volume:
        return 0  # Рішення недійсне, якщо перевищено об'єм
    return total_price  # Повертає сумарну вартість для дійсних рішень

3. Визначити початкову популяцію.

Кожен елемент популяції буде бінарним вектором, де 1 означає, що продукт включено до вантажівки, а 0 - що не включено.

4. Задати параметри генетичного алгоритму: кількість популяцій, кількість батьківських хромосом, що приймають участь у кросовері, тип кросовера та мутації тощо.

In [5]:
# Параметри генетичного алгоритму
num_generations = 100  # Кількість поколінь
num_parents_mating = 4  # Кількість батьків для кросоверу
sol_per_pop = 20  # Розмір популяції
num_genes = len(products)  # Кількість генів (товарів)

# Використання різних кросоверів для аналізу їхнього впливу
crossover_type = "single_point"  # Інший варіант - "two_points", "uniform"
mutation_type = "random"  # Мутація, яка змінює випадковий ген
mutation_probability = 0.1  # Ймовірність мутації

5. Створити екземпляр генеричного алгоритму засобами PyGad.

In [8]:
# Створення екземпляра генетичного алгоритму
ga_instance = pygad.GA(
    num_generations=num_generations,
    num_parents_mating=num_parents_mating,
    fitness_func=fitness_func,
    sol_per_pop=sol_per_pop,
    num_genes=num_genes,
    gene_type=int,
    init_range_low=0,
    init_range_high=2,
    parent_selection_type="sss",
    crossover_type=crossover_type,
    mutation_type=mutation_type,
    mutation_probability=mutation_probability
)

/usr/local/lib/python3.10/dist-packages/pygad/pygad.py:1139: UserWarning: The 'delay_after_gen' parameter is deprecated starting from PyGAD 3.3.0. To delay or pause the evolution after each generation, assign a callback function/method to the 'on_generation' parameter to adds some time delay.
  warnings.warn("The 'delay_after_gen' parameter is deprecated starting from PyGAD 3.3.0. To delay or pause the evolution after each generation, assign a callback function/method to the 'on_generation' parameter to adds some time delay.")


6. Дослідити вплив різних кросоверів та мутацій на результат.

In [9]:
# Запуск алгоритму
ga_instance.run()

# Отримання найкращого рішення
solution, solution_fitness, solution_idx = ga_instance.best_solution()
print("Найкраще рішення (вибір товарів):", solution)
print("Загальна вартість:", solution_fitness)
print("Загальний об'єм:", np.sum(solution * volumes))

Найкраще рішення (вибір товарів): [1 1 1 1 0 1 0 0 1 0 1 1]
Загальна вартість: 21002.35
Загальний об'єм: 2.5959000000000003


6. Дослідити вплив різних кросоверів та мутацій на результат.

In [10]:
# Гіперпараметри
num_generations = 50
sol_per_pop = 20
num_parents_mating = 10

# Типи кросоверів і діапазон ймовірностей мутації для тестування
crossover_types = ["single_point", "two_points", "uniform"]
mutation_probabilities = [0.1, 0.2, 0.3]

# Зберігаємо результати для аналізу
best_results = []

for crossover_type in crossover_types:
    for mutation_probability in mutation_probabilities:
        print(f"Testing crossover: {crossover_type}, mutation_probability: {mutation_probability}")

        # Створення екземпляра генетичного алгоритму
        ga_instance = pygad.GA(
            num_generations=num_generations,
            num_parents_mating=num_parents_mating,
            fitness_func=fitness_func,
            sol_per_pop=sol_per_pop,
            num_genes=num_genes,
            gene_type=int,
            init_range_low=0,
            init_range_high=2,
            parent_selection_type="sss",
            crossover_type=crossover_type,
            mutation_type="random",
            mutation_probability=mutation_probability
        )

        # Запуск оптимізації
        ga_instance.run()

        # Отримання кращого рішення та фітнесу
        solution, solution_fitness, _ = ga_instance.best_solution()
        best_results.append({
            "crossover_type": crossover_type,
            "mutation_probability": mutation_probability,
            "best_fitness": solution_fitness,
            "best_solution": solution
        })

        # Виведення результату для поточної комбінації параметрів
        print(f"Best fitness for {crossover_type} crossover and mutation {mutation_probability}: {solution_fitness}")

# Пошук найкращого рішення серед усіх параметрів
best_config = max(best_results, key=lambda x: x["best_fitness"])
print("\nBest configuration:")
print(f"Crossover Type: {best_config['crossover_type']}")
print(f"Mutation Probability: {best_config['mutation_probability']}")
print(f"Best Fitness: {best_config['best_fitness']}")
print(f"Best Solution: {best_config['best_solution']}")

Testing crossover: single_point, mutation_probability: 0.1
Best fitness for single_point crossover and mutation 0.1: 19231.64
Testing crossover: single_point, mutation_probability: 0.2
Best fitness for single_point crossover and mutation 0.2: 20693.69
Testing crossover: single_point, mutation_probability: 0.3
Best fitness for single_point crossover and mutation 0.3: 21773.77
Testing crossover: two_points, mutation_probability: 0.1
Best fitness for two_points crossover and mutation 0.1: 19783.239999999998
Testing crossover: two_points, mutation_probability: 0.2
Best fitness for two_points crossover and mutation 0.2: 20282.43
Testing crossover: two_points, mutation_probability: 0.3
Best fitness for two_points crossover and mutation 0.3: 19323.589999999997
Testing crossover: uniform, mutation_probability: 0.1
Best fitness for uniform crossover and mutation 0.1: 21731.54
Testing crossover: uniform, mutation_probability: 0.2
Best fitness for uniform crossover and mutation 0.2: 22082.43
Test

7. Зробити висновок про ефективність застосування генетичного алгоритму та оптимальний набір його параметрів.

**Оптимальний набір параметрів**
За результатами експериментів оптимальними налаштуваннями виявились:

**Тип кросоверу**: uniform (рівномірний)

**Ймовірність мутації**: 0.2

**Найкраще досягнуте значення фітнесу** (вартість обраних товарів): 22082.43

Це говорить про те, що рівномірний кросовер у поєднанні з середньою ймовірністю мутації (0.2) забезпечує оптимальний баланс між варіативністю рішень та здатністю алгоритму знаходити кращі розв'язки, не виходячи за межі об'єму вантажівки. Цей тип кросовера дозволяє комбінувати генетичну інформацію від батьків із більшою варіативністю, що знижує ймовірність застою в локальних максимумах.